File: mobile/composeApp/src/commonMain/kotlin/org/androidbible/data/sword/reader/ZTextReader.kt
package org.androidbible.data.sword.reader

import org.androidbible.data.sword.SwordModuleConfig
import org.androidbible.data.sword.SwordVersification
import org.androidbible.data.sword.io.BinaryFileReader
import org.androidbible.data.sword.io.CipherUtils
import org.androidbible.data.sword.io.ZlibDecompressor

class ZTextReader(
    private val config: SwordModuleConfig,
    private val modulePath: String,
) {
    private val blockCache = mutableMapOf<Pair<Int, Long>, ByteArray>()
    private val maxCacheEntries = 10

    fun readVerse(bookOsisId: String, chapter: Int, verse: Int): String {
        val (bookIndex, _) = SwordVersification.findBookByOsisId(bookOsisId)
            ?: return ""

        val testament = SwordVersification.getTestament(bookIndex)
        val testamentBookIndex = SwordVersification.getTestamentBookIndex(bookIndex)
        val books = if (testament == 0) SwordVersification.otBooks else SwordVersification.ntBooks

        val linearIndex = SwordVersification.computeLinearIndex(
            testamentBookIndex, chapter, verse, books
        )

        return readVerseByIndex(testament, linearIndex)
    }

    fun readChapter(bookOsisId: String, chapter: Int): List<Pair<Int, String>> {
        val (bookIndex, bookDef) = SwordVersification.findBookByOsisId(bookOsisId)
            ?: return emptyList()

        if (chapter < 1 || chapter > bookDef.chapterCount) return emptyList()

        val verseCount = SwordVersification.getVerseCount(bookIndex, chapter)
        val testament = SwordVersification.getTestament(bookIndex)
        val testamentBookIndex = SwordVersification.getTestamentBookIndex(bookIndex)
        val books = if (testament == 0) SwordVersification.otBooks else SwordVersification.ntBooks

        val result = mutableListOf<Pair<Int, String>>()

        for (v in 1..verseCount) {
            val linearIndex = SwordVersification.computeLinearIndex(
                testamentBookIndex, chapter, v, books
            )
            val text = readVerseByIndex(testament, linearIndex)
            if (text.isNotBlank()) {
                result.add(v to text)
            }
        }

        return result
    }

    private fun readVerseByIndex(testament: Int, linearIndex: Long): String {
        val prefix = if (testament == 0) "ot" else "nt"
        val dataDir = "$modulePath/${config.dataPath}"
        val bzvPath = "$dataDir/$prefix.bzv"
        val bzsPath = "$dataDir/$prefix.bzs"
        val bzzPath = "$dataDir/$prefix.bzz"

        val bzvReader = BinaryFileReader(bzvPath)
        try {
            val bzvOffset = linearIndex * 10
            if (bzvOffset + 10 > bzvReader.fileSize()) return ""

            val bzvEntry = bzvReader.readBytes(bzvOffset, 10)
            val blockNum = ByteUtils.readUInt32LE(bzvEntry, 0)
            val verseStart = ByteUtils.readUInt32LE(bzvEntry, 4)
            val verseSize = ByteUtils.readUInt16LE(bzvEntry, 8)

            if (verseSize == 0) return ""

            val block = getBlock(testament, blockNum, bzsPath, bzzPath)
                ?: return ""

            val start = verseStart.toInt()
            val end = minOf(start + verseSize, block.size)
            if (start >= block.size || start < 0) return ""

            val verseBytes = block.copyOfRange(start, end)
            val decrypted = CipherUtils.applyCipher(verseBytes, config)
            return decrypted.decodeToString()
        } finally {
            bzvReader.close()
        }
    }

    private fun getBlock(testament: Int, blockNum: Long, bzsPath: String, bzzPath: String): ByteArray? {
        val cacheKey = testament to blockNum

        blockCache[cacheKey]?.let { return it }

        val bzsReader = BinaryFileReader(bzsPath)
        try {
            val bzsOffset = blockNum * 12
            if (bzsOffset + 12 > bzsReader.fileSize()) return null

            val bzsEntry = bzsReader.readBytes(bzsOffset, 12)
            val compOffset = ByteUtils.readUInt32LE(bzsEntry, 0)
            val compSize = ByteUtils.readUInt32LE(bzsEntry, 4)
            val uncompSize = ByteUtils.readUInt32LE(bzsEntry, 8)

            if (compSize == 0L || uncompSize == 0L) return null

            val bzzReader = BinaryFileReader(bzzPath)
            try {
                val compressed = bzzReader.readBytes(compOffset, compSize.toInt())
                val decompressed = ZlibDecompressor.decompress(compressed, uncompSize.toInt())

                if (blockCache.size >= maxCacheEntries) {
                    blockCache.remove(blockCache.keys.first())
                }
                blockCache[cacheKey] = decompressed

                return decompressed
            } finally {
                bzzReader.close()
            }
        } finally {
            bzsReader.close()
        }
    }

    fun clearCache() {
        blockCache.clear()
    }
}

File: backend/app/Services/Sword/Readers/ZTextReader.php
<?php

namespace App\Services\Sword\Readers;

use App\Services\Sword\Versification\KjvVersification;
use App\Services\Sword\Versification\VersificationInterface;

/**
 * Reads SWORD zText (compressed text) modules.
 *
 * Binary file format for zText:
 *   - .bzs: Block start index (12 bytes/entry: uint32 offset, uint32 comp_size, uint32 uncomp_size)
 *   - .bzv: Verse index (10 bytes/entry: uint32 block_num, uint32 offset_in_block, uint16 text_size)
 *   - .bzz: Compressed data blocks (zlib deflated)
 *
 * Separate files exist for OT (ot.bzs, ot.bzv, ot.bzz) and NT (nt.bzs, nt.bzv, nt.bzz).
 *
 * To read a verse:
 * 1. Compute the flat index for the verse using the versification system
 * 2. Read the .bzv entry at (index * 10) to get block_num, offset_in_block, text_size
 * 3. Read the .bzs entry at (block_num * 12) to get block_offset, compressed_size, uncompressed_size
 * 4. Read compressed_size bytes from .bzz at block_offset
 * 5. Decompress with zlib
 * 6. Extract text_size bytes starting at offset_in_block from the decompressed data
 */
class ZTextReader implements ReaderInterface
{
    private VersificationInterface $versification;
    private string $dataPath;
    private ?string $cipherKey;

    /**
     * Cache of decompressed blocks to avoid re-reading during bulk operations.
     * Keyed by "{testament}:{block_num}".
     *
     * @var array<string, string>
     */
    private array $blockCache = [];

    public function __construct(string $dataPath, ?string $cipherKey = null, ?VersificationInterface $versification = null)
    {
        $this->dataPath = rtrim($dataPath, '/\\');
        $this->cipherKey = $cipherKey;
        $this->versification = $versification ?? new KjvVersification();
    }

    public function getDriverName(): string
    {
        return 'zText';
    }

    public function readVerse(string $osisId, int $chapter, int $verse): ?string
    {
        $testament = $this->versification->getTestament($osisId);
        $flatIndex = $this->versification->flatIndex($osisId, $chapter, $verse);

        return $this->readByIndex($testament, $flatIndex);
    }

    public function readChapter(string $osisId, int $chapter): array
    {
        $testament = $this->versification->getTestament($osisId);
        $verseCount = $this->versification->getVerseCount($osisId, $chapter);
        $verses = [];

        for ($v = 1; $v <= $verseCount; $v++) {
            $flatIndex = $this->versification->flatIndex($osisId, $chapter, $v);
            $text = $this->readByIndex($testament, $flatIndex);
            if ($text !== null && $text !== '') {
                $verses[$v] = $text;
            }
        }

        return $verses;
    }

    /**
     * Read all chapters for a book (used during import).
     *
     * @return array<int, array<int, string>> chapter => [verse => text]
     */
    public function readBook(string $osisId): array
    {
        $chapterCount = $this->versification->getChapterCount($osisId);
        $chapters = [];

        for ($c = 1; $c <= $chapterCount; $c++) {
            $chapterVerses = $this->readChapter($osisId, $c);
            if (!empty($chapterVerses)) {
                $chapters[$c] = $chapterVerses;
            }
        }

        // Clear block cache after processing a whole book
        $this->blockCache = [];

        return $chapters;
    }

    /**
     * Read verse text by flat index.
     */
    private function readByIndex(string $testament, int $flatIndex): ?string
    {
        $bzvPath = $this->dataPath . DIRECTORY_SEPARATOR . $testament . '.bzv';
        $bzsPath = $this->dataPath . DIRECTORY_SEPARATOR . $testament . '.bzs';
        $bzzPath = $this->dataPath . DIRECTORY_SEPARATOR . $testament . '.bzz';

        if (!file_exists($bzvPath) || !file_exists($bzsPath) || !file_exists($bzzPath)) {
            return null;
        }

        // Read the verse index entry (10 bytes)
        $bzvEntry = $this->readBytes($bzvPath, $flatIndex * 10, 10);
        if ($bzvEntry === null || strlen($bzvEntry) < 10) {
            return null;
        }

        $verseData = unpack('VblockNum/VoffsetInBlock/vtextSize', $bzvEntry);
        if (!$verseData || $verseData['textSize'] === 0) {
            return null;
        }

        $blockNum = $verseData['blockNum'];
        $offsetInBlock = $verseData['offsetInBlock'];
        $textSize = $verseData['textSize'];

        // Get decompressed block data
        $cacheKey = "{$testament}:{$blockNum}";
        if (!isset($this->blockCache[$cacheKey])) {
            $blockData = $this->decompressBlock($bzsPath, $bzzPath, $blockNum);
            if ($blockData === null) {
                return null;
            }
            $this->blockCache[$cacheKey] = $blockData;

            // Keep cache bounded (max 20 blocks)
            if (count($this->blockCache) > 20) {
                $oldest = array_key_first($this->blockCache);
                unset($this->blockCache[$oldest]);
            }
        }

        $blockData = $this->blockCache[$cacheKey];

        // Extract verse text from decompressed block
        if ($offsetInBlock + $textSize > strlen($blockData)) {
            return null;
        }

        $text = substr($blockData, $offsetInBlock, $textSize);

        // Apply cipher key if present (XOR)
        if ($this->cipherKey !== null) {
            $text = $this->applyCipher($text);
        }

        // Remove null bytes
        $text = rtrim($text, "\0");

        return $text ?: null;
    }

    /**
     * Decompress a data block from the .bzz file.
     */
    private function decompressBlock(string $bzsPath, string $bzzPath, int $blockNum): ?string
    {
        // Read block start entry (12 bytes)
        $bzsEntry = $this->readBytes($bzsPath, $blockNum * 12, 12);
        if ($bzsEntry === null || strlen($bzsEntry) < 12) {
            return null;
        }

        $blockInfo = unpack('VblockOffset/VcompSize/VuncompSize', $bzsEntry);
        if (!$blockInfo || $blockInfo['compSize'] === 0) {
            return null;
        }

        // Read compressed data
        $compressed = $this->readBytes($bzzPath, $blockInfo['blockOffset'], $blockInfo['compSize']);
        if ($compressed === null) {
            return null;
        }

        // Decompress (ZIP = zlib)
        $decompressed = @gzuncompress($compressed);
        if ($decompressed === false) {
            // Try raw deflate
            $decompressed = @gzinflate($compressed);
            if ($decompressed === false) {
                return null;
            }
        }

        return $decompressed;
    }

    /**
     * Read bytes from a file at a given offset.
     */
    private function readBytes(string $path, int $offset, int $length): ?string
    {
        $fh = @fopen($path, 'rb');
        if (!$fh) {
            return null;
        }

        fseek($fh, $offset);
        $data = fread($fh, $length);
        fclose($fh);

        return $data !== false ? $data : null;
    }

    /**
     * Apply XOR cipher (for locked modules).
     */
    private function applyCipher(string $data): string
    {
        $keyLen = strlen($this->cipherKey);
        $result = '';
        for ($i = 0; $i < strlen($data); $i++) {
            $result .= chr(ord($data[$i]) ^ ord($this->cipherKey[$i % $keyLen]));
        }
        return $result;
    }
}
